clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/IngestionBatchProcessor.java

```diff
@@ -71,32 +72,41 @@ public IngestionBatchProcessor(
     this.version = Version.parseVersionFromKafkaTopicName(storeVersionName);
   }
 
+  // For testing
+  KeyLevelLocksManager getLockManager() {
+    return this.lockManager;
+  }
+
   /**
    * When {@link #lockManager} is not null, this function will try to lock all the keys
    * (except Control Messages) passed by the params.
    */
-  public List<ReentrantLock> lockKeys(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records) {
+  public Map<ByteArrayKey, ReentrantLock> lockKeys(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records) {
     if (lockManager != null) {
-      List<ReentrantLock> locks = new ArrayList<>(records.size());
+      /**
+       * Need to use a {@link TreeMap} to make sure the locking will be executed in a deterministic order, otherwise
+       * deadlock can happen.
+       * Considering there could be multiple consumers, which are executing this function concurrently, and if they
+       * are trying to lock the same set of keys with different orders, deadlock can happen.
+       */
+      Map<ByteArrayKey, ReentrantLock> keyLockMap =
+          new TreeMap<>((o1, o2) -> ByteUtils.compare(o1.getContent(), o2.getContent()));
```

Comment:
can we compare with the `ByteArrayKey::hashCode()` should be faster than comparing byte arrays.

Reply:
Hash collision can become a disaster, which might result in deadlock.

Reply:
Not sure where the current ByteArrayKey hashCode is being used, but current hash generation looks too simple, does not use SHA or MD5. It might use hash collision more frequently.